ree/reeResidualFunctionals.cc

```diff
@@ -0,0 +1,923 @@
+    apf::MeshEntity* firstTetOppVert = getTetOppVert(
+        p->mesh, firstTet, firstFace);
+    apf::Vector3 firstTetOppVertCrd;
+    p->mesh->getPoint(firstTetOppVert, 0, firstTetOppVertCrd);
+
+    // last tet opp vertex crd
+    apf::MeshEntity* lastTetOppVert = getTetOppVert(
+        p->mesh, lastTet, firstFace);
+    apf::Vector3 lastTetOppVertCrd;
+    p->mesh->getPoint(lastTetOppVert, 0, lastTetOppVertCrd);
+
+    // normal to the face
+    apf::Downward edge_vertices;
+    p->mesh->getDownward(p->entity, 0, edge_vertices);
+    apf::Vector3 p0, p1, p2;
+    p->mesh->getPoint(edge_vertices[0], 0, p0);
+    p->mesh->getPoint(edge_vertices[1], 0, p1);
+    apf::MeshEntity* faceOppVert = getFaceOppVert(
+        p->mesh, firstFace, p->entity);
+    p->mesh->getPoint(faceOppVert, 0, p2);
+
+    apf::Vector3 normal = apf::cross(p1-p0, p2-p0);
+
+    // direction vectors from p0 to opp tet verts
+    apf::Vector3 vFirst = firstTetOppVertCrd - p0;
+    apf::Vector3 vLast = lastTetOppVertCrd - p0;
+
+    if ((vFirst * normal > 0)  && (vLast * normal < 0)) {
+      // reverse list of tets and faces
+      std::reverse(p->tets.begin(), p->tets.end());
+      std::reverse(p->faces.begin(), p->faces.begin());
+    }
+  }
+}
+
+static void runErm(EdgePatch* ep)
+{
+  getOrderedTetsandFaces(ep->mesh, ep->entity, ep->tets, ep->faces);
+  //getClockwiseTetsandFaces(ep);
+  assembleEdgePatchLHS(ep);
+  assembleEdgePatchRHS(ep);
+  mth::solveFromQR(ep->qr.Q, ep->qr.R, ep->b, ep->x);
+
+  if (!ep->isOnBdry) { // solve At*mu = g for g
+    mth::Vector<double> temp(ep->tets.size());
+    mth::multiply(ep->At, ep->x, temp);
+    for (size_t i = 0; i < ep->tets.size(); i++) {
+      ep->x(i) = temp(i);
+    }
+  }
+
+  int nf = ep->faces.size();
+  for(int i = 0; i < nf; i++) {
+    apf::MeshEntity* face = ep->faces[i];
+
+    apf::Downward e;
+    int ned = ep->mesh->getDownward(face, 1, e);
+    int ei = apf::findIn(e, ned, ep->entity);
+    PCU_ALWAYS_ASSERT(ned == 3 && ei != -1);
+
+    double components[3];
+    apf::getComponents(ep->equilibration->g, face, 0, components);
+    components[ei] = ep->x(i);
+    apf::setComponents(ep->equilibration->g, face, 0, components);
+  }
+}
+
+
+class EdgePatchOp : public apf::CavityOp
+{
+public:
+  EdgePatchOp(Equilibration* eq):
+    apf::CavityOp(eq->mesh)
+  {
+    setupEdgePatch(&edgePatch, eq);
+  }
+  virtual Outcome setEntity(apf::MeshEntity* e)
+  {
+    if (edgePatch.mesh->hasTag(e, edgePatch.equilibration->tag))
+      return SKIP;
+    startEdgePatch(&edgePatch, e);
+    if ( ! buildEdgePatch(&edgePatch, this))
+      return REQUEST;
+    return OK;
+  }
+  virtual void apply()
+  {
+    runErm(&edgePatch);
+    int n = VISITED;
+    edgePatch.mesh->setIntTag(
+        edgePatch.entity, edgePatch.equilibration->tag, &n);
+  }
+  EdgePatch edgePatch;
+};
+
+apf::Field* equilibrateResiduals(apf::Field* f)
+{
+  apf::Field* g = createPackedField(
+      apf::getMesh(f), "g", 3, apf::getConstant(2) );
```

Comment:
Is the `3` here because a 3d tet mesh is assumed?

Reply:
I needed to store 3 values on each face. So, that is why components=3 is used here.